File: Lista1/classes/Point3D/Point3D.cpp
#include "Point3D.h"
#include "../Vector3D/Vector3D.h"

// Default constructor for point 3D
Point3D::Point3D() {}

// Create point 3D
Point3D::Point3D(int posX, int posY, int posZ) : Point2D(posX, posY) {
    z = posZ;
}

// Return position Z of point 3D
int Point3D::getZ() {
    return z;
}

// Print point 3D information
void Point3D::printPoint() {
    std::cout << "Point 3D" << std::endl;
    std::cout << "X = " << getX() << std::endl;
    std::cout << "Y = " << getY() << std::endl;
    std::cout << "Z = " << getZ() << std::endl;
}

// Projection on point 2D from point 3D
Point2D Point3D::projectingToPoint2D(int z0, int d) {
    if ((z0 - getZ()) == 0) return *new Point2D(0, 0);
    int x2 = (z0 * getX() - getZ() * d) / (z0 - getZ());
    int y2 = (z0 * getY()) / (z0 - getZ());
    return *new Point2D(x2, y2);
}

// Move point by vector 3D
void Point3D::moveByVector(Vector3D vectorToMoveBy) {
    x += vectorToMoveBy.getX();
    y += vectorToMoveBy.getY();
    z += vectorToMoveBy.getZ();
}

File: Lista1Java/src/Point3d.java
/**
 * Class for creating 3D points
 */
public class Point3d extends Point2d {
    /** Position Z */
    protected int z;

    public Point3d(int posX, int posY, int posZ) {
        super(posX, posY);
        this.z = posZ;
    }

    /**
     * Return position Z value
     * @return int
     */
    public int getZ() {
        return this.z;
    }

    /**
     * Print some information about Point 3D
     */
    public void printPoint() {
        System.out.print("Point 3D\n");
        System.out.print("X = " + this.x + "\n");
        System.out.print("Y = " + this.y + "\n");
        System.out.print("Z = " + this.z + "\n");
    }

    /**
     * Projection on point 2D from point 3D
     * @param z0 parameter 1
     * @param d  parameter 2
     * @return   Point2d
     */
    public Point2d projectingToPoint2d(int z0, int d) {
        if ((z0 - this.z) == 0) return new Point2d(0, 0);
        int x2 = (z0 * this.x - this.z * d) / (z0 - this.z);
        int y2 = (z0 * this.y) / (z0 - this.z);
        return new Point2d(x2, y2);
    }

    /**
     * Move point by vector 3D
     * @param vectorToMoveBy vector to move by
     */
    public void moveByVector(Vector3d vectorToMoveBy) {
        this.x += vectorToMoveBy.getX();
        this.y += vectorToMoveBy.getY();
        this.z += vectorToMoveBy.getZ();
    }
}
